app/src/main/kotlin/com/simplemobiletools/camera/implementations/CameraXPreview.kt

```diff
@@ -0,0 +1,482 @@
+import androidx.camera.video.Recorder
+import androidx.camera.video.Recording
+import androidx.camera.video.VideoCapture
+import androidx.camera.video.VideoRecordEvent
+import androidx.camera.view.PreviewView
+import androidx.core.view.doOnLayout
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.window.layout.WindowMetricsCalculator
+import com.bumptech.glide.load.ImageHeaderParser.UNKNOWN_ORIENTATION
+import com.simplemobiletools.camera.R
+import com.simplemobiletools.camera.extensions.config
+import com.simplemobiletools.camera.extensions.toAppFlashMode
+import com.simplemobiletools.camera.extensions.toCameraXFlashMode
+import com.simplemobiletools.camera.helpers.MediaSoundHelper
+import com.simplemobiletools.camera.helpers.PinchToZoomOnScaleGestureListener
+import com.simplemobiletools.camera.helpers.ZoomCalculator
+import com.simplemobiletools.camera.interfaces.MyPreview
+import com.simplemobiletools.commons.extensions.showErrorToast
+import com.simplemobiletools.commons.extensions.toast
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+import kotlin.math.abs
+import kotlin.math.max
+import kotlin.math.min
+
+class CameraXPreview(
+    private val activity: AppCompatActivity,
+    private val previewView: PreviewView,
+    private val listener: CameraXPreviewListener,
+) : MyPreview, DefaultLifecycleObserver {
+
+    companion object {
+        private const val TAG = "CameraXPreview"
+        private const val RATIO_4_3_VALUE = 4.0 / 3.0
+        private const val RATIO_16_9_VALUE = 16.0 / 9.0
+
+        // Auto focus is 1/6 of the area.
+        private const val AF_SIZE = 1.0f / 6.0f
+        private const val AE_SIZE = AF_SIZE * 1.5f
+    }
+
+    private val config = activity.config
+    private val contentResolver = activity.contentResolver
+    private val mainExecutor = activity.mainExecutor
+    private val displayManager = activity.getSystemService(Context.DISPLAY_SERVICE) as DisplayManager
+    private val mediaSoundHelper = MediaSoundHelper()
+    private val windowMetricsCalculator = WindowMetricsCalculator.getOrCreate()
+
+    private val orientationEventListener = object : OrientationEventListener(activity, SensorManager.SENSOR_DELAY_NORMAL) {
+        @SuppressLint("RestrictedApi")
+        override fun onOrientationChanged(orientation: Int) {
+            if (orientation == UNKNOWN_ORIENTATION) {
+                return
+            }
+
+            val rotation = when (orientation) {
+                in 45 until 135 -> Surface.ROTATION_270
+                in 135 until 225 -> Surface.ROTATION_180
+                in 225 until 315 -> Surface.ROTATION_90
+                else -> Surface.ROTATION_0
+            }
+
+            preview?.targetRotation = rotation
+            imageCapture?.targetRotation = rotation
+            videoCapture?.targetRotation = rotation
+        }
+    }
+
+    private val hasBackCamera: Boolean
```

Comment:
lets use functions for these please, I never liked the way these "get()"s look and work

Reply:
Okay. Will make the change.

Reply:
Done